File: quad.c
#include <stdio.h>
#include <stdbool.h>
#include <time.h>


struct Square {
    int x1, y1, x2, y2, d, pos;
    struct Square* topLeft;
    struct Sqaure* topRight;
    struct Square* botLeft;
    struct Square* botRight;
    bool leaf;
};

struct Square* initSquare(int x1, int y1, int x2, int y2, int d, int n, int pos) {
    
    struct Square* sq = (struct Square *) malloc(sizeof(struct Square));
    sq->x1 = x1;
    sq->y1 = y1;
    sq->x2 = x2;
    sq->y2 = y2;
    sq->d = d;
    sq->pos = pos;
    if (n >= d) {
        sq->leaf = false;
        initGrid( sq, n );
    } else {
        sq->leaf = true;
    }
    return sq;
}

void initGrid(struct Square* sq, int n) {

    int w = sq->x2 - sq->x1;
    int h = sq->y2 - sq->y1;
    int midX = sq->x1 + w/2;
    int midY = sq->y1 + h/2;

    sq->topLeft = initSquare( sq->x1, sq->y1, midX, midY, sq->d + 1, n, 1 );
    sq->topRight = initSquare( midX, sq->y1, sq->x2, midY, sq->d + 1, n, 2 );
    sq->botLeft = initSquare( sq->x1, midY, midX, sq->y2, sq->d + 1, n, 3 );
    sq->botRight = initSquare(midX, midY, sq->x2, sq->y2, sq->d + 1, n, 4);

}


int getPosFromPoint(struct Square* sq, int x, int y) {
    int w = sq->x2 - sq->x1;
    int h = sq->y2 - sq->y1;
    int midX = sq->x1 + w/2;
    int midY = sq->y1 + h/2;
    bool isTop = false;
    bool isLeft = false;

    if ( x < midX ) isLeft = true;
    if ( y < midY ) isTop = true;

    if ( isTop && isLeft ) return 1;
    else if ( isTop && !isLeft ) return 2;
    else if ( !isTop && isLeft ) return 3;
    else if ( !isTop && !isLeft) return 4;
}

int getPoint(struct Square* sq, int x, int y, int index) {
    if(sq->leaf) return index;
    int pos = getPosFromPoint(sq, x, y);

    switch(pos) {
        case 1:
            // 00
            printf("topLeft\n");
            getPoint(sq->topLeft, x, y, index);
            break;
        case 2:
            // 01
            printf("topRight\n");
            index = index | (1 << sq->d*2);
            getPoint(sq->topRight, x, y, index);
            break;
        case 3:
            // 10
            printf("botLeft\n");
            index = index | (2 << sq->d*2);
            getPoint(sq->botLeft, x, y, index);
            break;
        case 4:
            // 11
            printf("botRight\n");
            index = index | (3 << sq->d*2);
            getPoint(sq->botRight, x, y, index);
            break;
    }    
}

int getRange( struct Square* sq, int x1, int x2, int y1, int y2) {
    if( sq->leaf ) return 1;

    int pos1 = getPosFromPoint(sq, x1, y1);
    int pos2 = getPosFromPoint(sq, x2, y2);
    int pos3 = getPosFromPoint(sq, x1, y2);
    int pos4 = getPosFromPoint(sq, x2, y1);

    recurseByPos( sq, pos1, x1, y1, x2, y2 );
    recurseByPos( sq, pos2, x1, y1, x2, y2 );
    recurseByPos( sq, pos3, x1, y1, x2, y2 );
    recurseByPos( sq, pos4, x1, y1, x2, y2 );

}

void recurseByPos( struct Square* sq, int pos, int x1, int x2, int y1, int y2 ) {
    switch( pos ) {
        case 1:
            getRange( sq->topLeft, x1, y1, x2, y2);
            break;
        case 2:
            getRange( sq->topRight, x1, y1, x2, y2);
            break;
        case 3:
            getRange( sq->botLeft, x1, y1, x2, y2);
            break;
        case 4:
            getRange( sq->botRight, x1, y1, x2, y2);
            break;
    }
}

int main() {
   struct Square* sq = initSquare( 0, 0, 10000, 10000, 1, 6, 1 );
   int index = getPoint( sq, 9000, 9000, 0);
   printf("Result: %i\n", index);
   clock_t begin = clock();
   getRange( sq, 0, 0, 3000, 6000 );
   clock_t end = clock();
    double time_spent = (double)(end - begin) / CLOCKS_PER_SEC;
    printf("%f", time_spent);
   return 0;
}

File: quad.py
import sys, os
import time

class Square:

    def __init__(self, x1, y1, x2, y2, d, n, pos):
        self.x1 = x1
        self.y1 = y1
        self.x2 = x2
        self.y2 = y2
        self.d = d
        self.leaf = False
        self.pos = pos
        
        if (n >= self.d):
            self.initGrid(n)
        else:
            self.leaf = True

    def initGrid(self, n):
        w = self.x2 - self.x1
        h = self.y2 - self.y1
        midX = self.x1 + w/2
        midY = self.y1 + h/2
        self.topLeft = Square(self.x1, self.y1, midX, midY, self.d + 1, n, "topLeft")
        self.topRight = Square(midX, midY, self.x2, self.y2, self.d + 1, n, "topRight")
        self.botLeft = Square(self.x1, midY, midX, self.y2, self.d + 1, n, "botLeft")
        self.botRight = Square(midX, midY, self.x2, self.y2, self.d + 1, n, "botRight")

    def getPoint(self, square, x, y):
        #print(square.pos)
        if( square.leaf ):
            return square
        pos = self.getPosFromPoint(square, x, y)
        if ( pos == "topLeft" ):
            self.getPoint(square.topLeft, x, y)
        elif ( pos == "topRight" ):
            self.getPoint(square.topRight, x, y)
        elif ( pos == "botLeft" ):
            self.getPoint(square.botLeft, x, y)
        elif ( pos == "botRight" ):
            self.getPoint(square.botRight, x, y)

    def getPosFromPoint(self, square, x, y):
        w = square.x2 - square.x1
        h = square.y2 - square.y1
        midX = square.x1 + w/2
        midY = square.y1 + h/2

        isTop = False
        isLeft = False
        if ( x < midX ):
            isLeft = True
        if ( y < midY ):
            isTop = True

        if ( isTop and isLeft ):
            return "topLeft"
        elif ( isTop and not isLeft ):
            return "topRight"
        elif ( not isTop and isLeft ):
            return "botLeft"
        elif ( not isTop and not isLeft):
            return "botRight"   

    def getRange( self, square, x1, x2, y1, y2 ):

        if( square.leaf ):
            return square
        pos1 = self.getPosFromPoint(square, x1, y1)
        pos2 = self.getPosFromPoint(square, x2, y2)
        pos3 = self.getPosFromPoint(square, x1, y2)
        pos4 = self.getPosFromPoint(square, x2, y1)

        self.recurseByPos( square, pos1, x1, y1, x2, y2 )
        self.recurseByPos( square, pos2, x1, y1, x2, y2 )
        self.recurseByPos( square, pos3, x1, y1, x2, y2 )
        self.recurseByPos( square, pos4, x1, y1, x2, y2 )



    def recurseByPos( self, square, pos, x1, x2, y1, y2 ):
        if ( pos == "topLeft" ):
            self.getRange(square.topLeft, x1, y1, x2, y2)
        elif ( pos == "topRight" ):
            self.getRange(square.topRight, x1, y1, x2, y2)
        elif ( pos == "botLeft" ):
            self.getRange(square.botLeft, x1, y1, x2, y2)
        elif ( pos == "botRight" ):
            self.getRange(square.botRight, x1, y1, x2, y2)
        


square = Square( 0, 0, 10000, 10000, 1, 7, "Root")
#grid = square.getPoint(square, 7762, 3122)

tic = time.perf_counter()
square.getRange( square, 0, 0, 1920, 1080 )
toc = time.perf_counter()
print(f"{toc - tic:0.8f} seconds")
